### PR TITLE
Add instructions for pushing to GitHub from the VEDA Hub

### PR DESCRIPTION
**Pull request by maxrjones**
This PR adds a new section to `Usage Examples` called `JupyterHub`. It currently contains one markdown file `How to push to GitHub from the VEDA JupyterHub` but I think the new section would be helpful for other upcoming improvements (e.g., additional IDEs). While there's some redundancy with the 2i2c docs, I think the 2i2c docs offer cover background and reasons for things being done a certain way (which is useful for their audience), whereas these how-to guides would provide the minimal required explanation for solving a user's problem.

### CODE REVIEW

File: notebooks/quickstarts/github-authentication.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+1. Navigate to [https://github.com/apps/veda-hub-github-scoped-creds](https://github.com/apps/veda-hub-github-scoped-creds).
+2. Select `Configure`.
+3. Select the organization that contains the repository that you would like to push to.
+4. Select either `All repositories` or `Only select repositories` depending on whether you only want to allow pushing to specific repositories. In order for a Hub user to write to those repositories, the repository needs to be added to the application **and** the user needs to have read/write permissions for the repository. The following restrictions define whether you can install the application yourself or if you need to request permission:
+    - Organization owners can install GitHub Apps for all repositories.
+    - Repository admins can install the application if they only grant access to repositories that they administer.
+    - If a requested does not have sufficient permissions, GitHub will send a notification to the organization owner requesting that they install the app.
```
Comment (jsignell) [inline notebooks/quickstarts/github-authentication.md:15]:
```suggestion
    - If you do not have sufficient permissions, GitHub will send a notification to the organization owner requesting that they install the app.
```

**Review by jsignell — COMMENTED** [review]
This looks good, but I did get an error. Probably something that I had set up already that conflicts:

```
warning: credential.https://github.com.helper has multiple values
error: cannot overwrite multiple values with a single value
       Use a regexp, --add or --replace-all to change credential.https://github.com.helper.
```
Full error:

image

**Comment by maxrjones** [comment]
Thanks for the review @jsignell! @wildintellect mentioned that this setup may still be confusing to users, so I'm going to try out recording a quick video explanation (hopefully tomorrow). So, I'll hold off on merging this for a short while and may ping you for another review.

that error does seem likely relating to an existing setup. I can add a warning and perhaps we should also include a section in the JupyterHub section of the docs on where users can go to reach out for help. Is the IMPACT slack where that help would happen?

**Comment by wildintellect** [comment]
Step 1. Clarify that an Org admin has to setup the initial permissions.
Step 2.
> Second, you need to authorize the app on the Hub. This needs to be done once every 8 hours. 
Does every user need to do this? So basically it's once per working day?
Can we include `gh-scoped-creds.ipynb` in the docs repo, so that users just need to open it and run it? (we already checkout the docs in Pangeo workspaces)

The whole thing feels reverse order for actual user usage.
1. Remember to clone your repos with https, NOTE: github token is the alternative to this method
2. Authorize once per day (Every 8 hours) <- I wonder if we could make a widget or if the jupyter github extension incorporates this
3. First time Config for Repos/Organizations (maybe mention at top and link to bottom?

**Review by wildintellect — CHANGES_REQUESTED** [review]
See previous comment thread.

**Comment by wildintellect** [comment]
@maxrjones looks close I think the only thing missing is a copy of that notebook in the PR that goes into notebooks/templates ? open to other location ideas. I just don't want it to only exist in the shared folder on VEDA where it can go missing.

**Comment by maxrjones** [comment]
> @maxrjones looks close I think the only thing missing is a copy of that notebook in the PR that goes into notebooks/templates ? open to other location ideas. I just don't want it to only exist in the shared folder on VEDA where it can go missing.

It's just two lines of code which are shown in the GIF. Also, only admins should be able to modify the `shared-readwrite` directory, so I think the risk here is minimal.

Relatedly, could you please create that notebook in `shared-readwrite`? I would offer to become an admin, but I actually think it's useful for me to have the same experience as a typical VEDA user.

**Comment by maxrjones** [comment]
> > @maxrjones looks close I think the only thing missing is a copy of that notebook in the PR that goes into notebooks/templates ? open to other location ideas. I just don't want it to only exist in the shared folder on VEDA where it can go missing.
> 
> It's just two lines of code which are shown in the GIF. Also, only admins should be able to modify the `shared-readwrite` directory, so I think the risk here is minimal.
> 
> Relatedly, could you please create that notebook in `shared-readwrite`? I would offer to become an admin, but I actually think it's useful for me to have the same experience as a typical VEDA user.

@wildintellect do you think we could merge this PR? a Hub user is asking about sharing code and I think these instructions could be helpful

**Comment by wildintellect** [comment]
@maxrjones another reason to put it in the repo, this code will actually get checkout to instances and we won't need to manage `shared-public/shared-rewrite`.

**Comment by maxrjones** [comment]
> @maxrjones another reason to put it in the repo, this code will actually get checkout to instances and we won't need to manage `shared-public/shared-rewrite`.

OK, done

File: nasa-veda-platform/scientific-computing/github-authentication.qmd [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+Copy the `/home/jovyan/veda-docs/notebooks/templates/template-github-auth.ipynb` notebook to your personal home directory on the Hub as `github-auth.ipynb`.
+
+::: {.callout-tip}
+You only need to do this step once.
+:::
+
+### Authenticate to GitHub
+
+::: {.callout-tip}
+You need to do this step each time you create a new JupyterHub instance.
```
Comment (wildintellect) [inline nasa-veda-platform/scientific-computing/github-authentication.qmd:21]: Create or Boot (start)? I feel like the word could be confusing, but maybe I'm overthinking.

File: nasa-veda-platform/scientific-computing/github-authentication.qmd [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+:::
+
+::: {.callout-warning}
+This will only allow pushing to repositories that have been [configured with the veda-hub-github-scoped-creds-app]()
```
Comment (wildintellect) [inline nasa-veda-platform/scientific-computing/github-authentication.qmd:25]: Is this supposed to be a link to something?

Now I recall this step is the reason that I wanted to note Personal Access Tokens as a common alternative that doesn't require Org level admins to approve.

Comment (wildintellect) [reply nasa-veda-platform/scientific-computing/github-authentication.qmd:25]: It should say something like. This is the recommend method on VEDA, but if you're having issues you can fallback on this method. https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens that works everywhere.

**Review by wildintellect — APPROVED** [review]

**Pull request merged**